File: ms/sandbox/seir.cpp
#include <TMB.hpp>

template<class Type>
vector<Type> rungekutta(vector<Type> (*f)(Type,
                                          const vector<Type> &,
                                          const vector<Type> &,
                                          const vector<int> &),
                        matrix<Type> &F,
                        Type t,
                        const vector<Type> &y,
                        const vector<Type> &dtheta,
                        const vector<int> &itheta,
                        const vector<Type> &a,
                        const vector<Type> &b,
                        const vector<Type> &c,
                        Type h)
{
	Eigen::Index i, j, k, m = F.rows(), n = F.cols(), p = 0;
	vector<Type> ans(m);
	for (j = 0; j < n; ++j) {
		for (i = 0; i < m; ++i)
			ans(i) = y(i);
		for (k = 0; k < j; ++k, ++p)
		for (i = 0; i < m; ++i)
			ans(i) += a(p) * h * F(i, k);
		F.col(j) = f(t + c(j) * h, ans, dtheta, itheta);
	}
	for (i = 0; i < m; ++i)
		ans(i) = y(i);
	for (j = 0; j < n; ++j)
	for (i = 0; i < m; ++i)
		ans(i) += b(j) * h * F(i, j);
	return ans;
}

template<class Type>
vector<Type> dot(Type t,
                 const vector<Type> &y,
                 const vector<Type> &dtheta,
                 const vector<int> &itheta)
{
	int m = itheta(0);
	int n = itheta(1);
	Type s1 = (y.segment(1 + m, n)       ).exp().sum();
	Type s2 = (y.segment(1 + m, n) - y(1)).exp().sum();
	vector<Type> dydt(1 + m + n + 1 + 1);
	dydt(0) = dtheta(1) + dtheta(3 + m + n) * exp(y(1 + m + n)) -
		(dtheta(0) * s1 + dtheta(2)) * y(0);
	dydt(1) = dtheta(0) * s2 * y(0) - (dtheta(3) + dtheta(2));
	for (int k = 0; k < m + n; ++k)
	dydt(2 + k) = dtheta(3 + k) * exp(y(1 + k) - y(2 + k)) -
		(dtheta(4 + k) + dtheta(2));
	dydt(2 + m + n) = dtheta(0) * s1 * y(0);
	return dydt;
}

template<class Type>
Type objective_function<Type>::operator() ()
{
	/* .... Data .................................................... */

	/* Number of time points */
	DATA_INTEGER(T);

	/* Time series */
	DATA_IVECTOR(incidence);
	DATA_VECTOR(birth);
	DATA_VECTOR(death);

	/* Constants */
	DATA_SCALAR(intercept);
	DATA_SCALAR(sigma);
	DATA_SCALAR(gamma);
	DATA_SCALAR(delta);

	/* Number of E[i], I[j] compartments */
	DATA_INTEGER(m);
	DATA_INTEGER(n);

	/* State (S, E, I, R) at first time point */
	DATA_VECTOR(init);

	/* Nullity, rank */
#if 0
	DATA_INTEGER(R0);
#endif
	DATA_INTEGER(R1);

	/* Model matrices */
	DATA_MATRIX(X0);
	DATA_MATRIX(X1);

	/* Runge-Kutta coefficients */
	DATA_VECTOR(rka);
	DATA_VECTOR(rkb);
	DATA_VECTOR(rkc);


	/* .... Transformed data ........................................ */
	vector<Type> dtheta(3 + m + n + 1);
	for (int i = 0; i < m; ++i)
		dtheta(3 + i + 0 + 0) = m * sigma;
	for (int j = 0; j < n; ++j)
		dtheta(3 + m + j + 0) = n * gamma;
	for (int k = 0; k < 1; ++k)
		dtheta(3 + m + n + k) = 1 * delta;
	vector<int> itheta(2);
	itheta(0) = m;
	itheta(1) = n;


	/* .... Parameters .............................................. */

	PARAMETER(log_sd);
	PARAMETER(log_size);
	PARAMETER_VECTOR(b0);
	PARAMETER_VECTOR(b1);


	/* .... Transformed parameters .................................. */

	/* Spline */
	vector<Type> log_trans = intercept + X0 * b0 + X1 * b1;

	/* State (S, log(E), log(I), log(R), cumulative incidence) */
	matrix<Type> state(1 + m + n + 1 + 1, T);

	state(0, 0) = init(0);
	for (int i = 0; i < m; ++i)
		state(1 + i + 0 + 0, 0) = log(init(1 + i + 0 + 0));
	for (int j = 0; j < n; ++j)
		state(1 + m + j + 0, 0) = log(init(1 + m + j + 0));
	for (int k = 0; k < 1; ++k)
		state(1 + m + n + k, 0) = log(init(1 + m + n + k));
	state(1 + m + n + 1, 0) = Type(0.0);

	matrix<Type> F(state.rows(), rkb.size());
	vector<Type> y = state.col(0);
	for (int t = 1; t < T; ++t) {
		dtheta(0) = exp(log_trans(t - 1));
		dtheta(1) = birth(t - 1);
		dtheta(2) = death(t - 1);
		y = rungekutta(dot, F, Type(0.0), y, dtheta, itheta,
		               rka, rkb, rkc, Type(1.0));
		state.col(t) = y;
	}

	REPORT(log_trans);
	REPORT(state);
	ADREPORT(log_trans);
	ADREPORT(state);


	/* .... Model ................................................... */
	Type ans = Type(0.0);

	Type sd = exp(log_sd);
	for (int j = 0; j < R1; ++j)
		ans -= dnorm(b1(j), Type(0.0), sd, 1);

#if 0
	Type size = exp(log_size);
#endif
	for (int t = 1; t < T; ++t) {
		Type log_mu = log(state(1 + m + n + 1, t - 0) -
		                  state(1 + m + n + 1, t - 1));
		ans -= dnbinom_robust(Type(incidence(t)),
		                      log_mu, Type(2.0) * log_mu - log_size, 1);
	}


	return ans;
}

File: ms/sandbox/seir.R
## .... Set up spline stuff ............................................

iperm <- function (p) { p[p] <- seq_along(p); p }

basis.dimension <- 8L # number of basis functions
basis.order <- 4L # number of polynomial coefficients
penalty.order <- 2L # order of difference penalty

stopifnot(basis.dimension >= basis.order,
          basis.dimension > penalty.order)

T <- 250L # number of time points
times <- seq.int(from = 0, by = 1, length.out = T)

K <- basis.dimension + basis.order # number of knots
K. <- K - 2L * (basis.order - 1L) # number of internal knots
b <- 0.001 * (times[length(times)] - times[1L])
d <- 1.002 * (times[length(times)] - times[1L])/(K. - 1L)
knots <- seq.int(from = times[1L] - b - (basis.order - 1L) * d,
                 by = d, length.out = K)

X <- splines::splineDesign(knots = knots, x = times, ord = basis.order)
F <- qr(X)
S <- crossprod(diff(diag(basis.dimension), differences = penalty.order) %*%
               backsolve(F$qr, diag(dim(F$qr)[2L]))[iperm(F$pivot), ])

stopifnot(identical(dim(X), c(T, basis.dimension)),
          identical(dim(S), c(basis.dimension, basis.dimension)))

e <- eigen(S, symmetric = TRUE)
j <- rep(c(TRUE, FALSE), c(basis.dimension - penalty.order, penalty.order))
QU <- qr.Q(F) %*% e$vectors

X. <- QU[, !j, drop = FALSE]
Z. <- QU[,  j, drop = FALSE] * rep(1/sqrt(e$values[j]), each = T)


## .... Set up incidence stuff .........................................

N <- 1e+6L
m <- 1L
n <- 1L

beta.0 <- 1e-5
beta.1 <- 1e-1
nu.0 <- 1e+3
mu.0 <- nu.0/N
sigma <- 0.5
gamma <- 0.5
delta <- 0

beta <- function (t) beta.0 * (1 + beta.1 * sinpi(t / 26))
nu   <- function (t) rep(nu.0, length(t))
mu   <- function (t) rep(mu.0, length(t))

set.seed(0L)
init <- fastbeta::seir.ee(beta.0, nu.0, mu.0, sigma, gamma, delta, m, n)
series <- fastbeta::seir(T + 1L, beta, nu, mu, sigma, gamma, delta, m, n,
                         init, epsilon = 0.002)[-1L, ]


## .... Set up data ....................................................

data <- list(T = T,
             incidence = as.integer(series[, 1L + m + n + 1L + 1L]),
             birth = as.double(series[, 1L + m + n + 1L + 2L]),
             death = mu(times),
             intercept = gamma/series[1L, 1L],
             sigma = sigma,
             gamma = gamma,
             delta = delta,
             m = m,
             n = n,
             init = as.double(series[1L, 1L:(1L + m + n + 1L)]),
             R0 = ncol(X.),
             R1 = ncol(Z.),
             X0 = X.,
             X1 = Z.,
             rka = c(     1.0/     5.0,
                          3.0/    40.0,
                          9.0/    40.0,
                         44.0/    45.0,
                        -56.0/    15.0,
                         32.0/     9.0,
                      19372.0/  6561.0,
                     -25360.0/  2187.0,
                      64448.0/  6561.0,
                       -212.0/   729.0,
                       9017.0/  3168.0,
                       -355.0/    33.0,
                      46732.0/  5247.0,
                         49.0/   176.0,
                      -5103.0/ 18656.0,
                         35.0/   384.0,
                          0.0/     1.0,
                        500.0/  1113.0,
                        125.0/   192.0,
                      -2187.0/  6784.0,
                         11.0/    84.0),
             rkb = c(  5179.0/ 57600.0,
                          0.0/     1.0,
                       7571.0/ 16695.0,
                        393.0/   640.0,
                     -92097.0/339200.0,
                        187.0/  2100.0,
                          1.0/    40.0),
             rkc = c(     0.0/     1.0,
                          1.0/     5.0,
                          3.0/    10.0,
                          4.0/     5.0,
                          8.0/     9.0,
                          1.0/     1.0,
                          1.0/     1.0),
             rhk = 1)
